src/commands/keygen.rs

```diff
@@ -0,0 +1,380 @@
+        if let Some((name, params)) = value.split_once(':') {
+            #[allow(clippy::single_match)]
+            match name {
+                "RSASHA256" => {
+                    let bits: u32 = params.parse().map_err(|err| {
+                        clap::Error::raw(
+                            clap::error::ErrorKind::InvalidValue,
+                            format!("invalid RSA key size '{params}': {err}"),
+                        )
+                    })?;
+                    return Ok(GenerateParams::RsaSha256 { bits });
+                }
+                _ => {}
+            }
+        }
+
+        Err(clap::Error::raw(
+            clap::error::ErrorKind::InvalidValue,
+            format!("unrecognized algorithm '{value}'"),
+        ))
+    }
+
+    pub fn execute(self, env: impl Env) -> Result<(), Error> {
+        let mut stdout = env.stdout();
+
+        let params = self.algorithm;
+
+        // The digest algorithm is selected based on the key algorithm.
+        let digest_alg = match params.algorithm() {
+            SecAlg::RSASHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP256SHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP384SHA384 => DigestAlg::SHA384,
+            SecAlg::ED25519 => DigestAlg::SHA256,
+            SecAlg::ED448 => DigestAlg::SHA256,
+            _ => unreachable!(),
+        };
+
+        // Generate the key.
+        // TODO: Attempt repeated generation to avoid key tag collisions.
+        let (secret_key, public_key) = common::generate(params)
+            .map_err(|err| format!("an implementation error occurred: {err}").into())
+            .context("generating a cryptographic keypair")?;
+        let flags = if self.make_ksk { 257 } else { 256 };
+        let public_key = Key::new(self.name.clone(), flags, public_key);
+        let digest = self
+            .make_ksk
+            .then(|| public_key.digest(digest_alg).unwrap());
+
+        // Open the appropriate files to write the key.
+        let base = format!(
+            "K{}+{:03}+{:05}",
+            self.name.fmt_with_dot(),
+            public_key.algorithm().to_int(),
+            public_key.key_tag()
+        );
+        // TODO: Adjust for how 'Env' mocks the current directory.
+        let mut secret_key_file = File::create_new(format!("{base}.private"))
+            .map_err(|err| format!("cannot create '{base}.private': {err}"))?;
+        let mut public_key_file = File::create_new(format!("{base}.key"))
+            .map_err(|err| format!("public key file '{base}.key' already existed: {err}"))?;
+        let mut digest_file = self
+            .make_ksk
+            .then(|| File::create_new(format!("{base}.ds")))
+            .transpose()
+            .map_err(|err| format!("digest file '{base}.ds' already existed: {err}"))?;
+
+        #[cfg(target_family = "unix")]
+        if self.create_symlinks {
+            if let Ok(metadata) = std::fs::symlink_metadata(".private") {
```

Comment:
I take it there's no way to just force create a symlink?

Reply:
Yeah, the underlying `symlink(2)` syscall doesn't support overwriting.